asmcomp/cmmgen.ml

```diff
@@ -224,6 +224,112 @@ let untag_int = function
   | Cop(Cor, [c; Cconst_int 1]) -> Cop(Casr, [c; Cconst_int 1])
   | c -> Cop(Casr, [c; Cconst_int 1])
 
+let int_comp_with_constant cmp arg n =
+  let default = Cop(Ccmpi cmp, [arg; Cconst_int n]) in
+  match arg with
+  | Cop(Cor, [Cop(Clsr, [arg; Cconst_int shift]); Cconst_int mask])
+  | Cop(Cor, [Cconst_int mask; Cop(Clsr, [arg; Cconst_int shift])]) ->
+      (* Optimize the pattern:
+           ( (a >> shift) | mask )  <  n
+         into
+           ( a & ( !mask << shift ) ) < (n & !mask) << shift
+         when
+           A: n | mask = n
+           B: (n & !mask) << shift >> shift = (n & !mask)
+           C: sign ((n & !mask) << shift) = sign n
+
+         shift, mask and n being constants, the remaining operations are:
+            a & and_mask < n'
+         which is one less operation
+
+         =======
+
+         For all comparison operator (<)
+
+                              ( (a >> shift) | mask )  <  n
+
+         iff
+
+                              ( (a >> shift) | mask )  <  (n | mask)
```

Comment:
`n = n | mask` is a precondition (A). Otherwise this is effectively completely wrong. The assumption is that this will often correspond to the tagging, hence the constraint will hold. But this is also checked.